### PR TITLE
Add batched smaps support to MRINufftAutograd : Address #258 

### PR DESCRIPTION
**Pull request by AsmaTANABEN**
This PR addresses : [#issue 258](https://github.com/mind-inria/mri-nufft/issues/258), which requests support for batched sensitivity maps (smaps) in the MRINufftAutograd module.

✅ **Fixes / Changes**: 

**-** **Introduced BatchedNufftAutograd**: A new torch.nn.Module extending MRINufftAutograd, allowing the NUFFT operator to handle batched inputs with corresponding batched smaps, while retaining autograd support.
**-** **Updated get_operator function** (mrinufft/operators/base.py): Added 'use_batched_mode' and 'batch_size' as parameters, to dynamically select between the wrappers MRINufftAutograd and the new BatchedNufftAutograd.
**- Extended testing** in test_autodiff.py: Update unit tests to validate the batched mode when use_batched_mode=True.

 **Impact/Use cases :** 

-  **To use the BatchedNufftAutograd wrapper** (when autograd is enabled with respect to data or trajectory), initialize the operator as follows:
`nufft = mrinufft.get_operator(
    backend_name="gpunufft" or "cufinufft", 
    wrt_data=True, 
    wrt_traj=False,
    use_batched_mode=True,
    batch_size=batch_size)(**kwargs)`
 Setting use_batched_mode=True along with batch_size ensures that BatchedNufftAutograd is used instead of the default MRINufftAutograd

**- Enables:**
Calling the NUFFT operator with both batched inputs and batched smaps:
`nufft.adj_op(batched_smaps, batched_kspace_data)
nufft.op(batched_smaps, batched_imgs)`
The output includes an additional batch dimension compared to the original MRINufftAutograd.

**- Supports:**
**Batched training** for deep learning models (e.g., using mono-contrast datasets)
(✅ Tested with my NCPDNet implementation.)
**Single-batch training with multi-contrast data**, where the batch dimension represents multiple contrast channels (e.g., mp2rage imaging data).

### CODE REVIEW

**Comment by paquiteau** [comment]
I like this very much ! Two (minors) points of concern however:
- I would rename "use_batched_mode" to something more  explicit like `paired_batch_mode` (we already have batch support)
- It seems that the commit history is a bit messed up (the modification in the io module should not be in this PR)

File: src/mrinufft/io/siemens.py
Comment (paquiteau) [inline src/mrinufft/io/siemens.py:1]: should probably not be in this PR

File: src/mrinufft/operators/autodiff.py [hunk 1]
```diff
@@ -152,3 +152,94 @@ def samples(self, value):
...
+    nufft : object
+        An instance of a standard NUFFT operator.
+    batch_size : int
+        Number of batches to process simultaneously.
+    **kwargs : dict, optional
+        Additional arguments.
+
+    Notes
+    -----
+    #TODO Future improvements may include support for varying trajectories across batches.
```
Comment (paquiteau) [inline src/mrinufft/operators/autodiff.py:173]: Don't put this in the Docstring (A comment is fine though)

Inside the note you can details the math if you wish, and provide an example of usage as well

File: src/mrinufft/operators/autodiff.py [hunk 2]
```diff
@@ -152,3 +152,94 @@ def samples(self, value):
...
+        self._check_input_shape(imgs=batched_smaps)
+        batched_imgs = []
+        for i in range(self.batch_size):
+            try:
+                self.nufft_op.smaps = batched_smaps[i]
+                batched_imgs.append(
+                    _NUFFT_ADJOP.apply(batched_kspace[i], self.samples, self.nufft_op)
+                )
+            except Exception as e:
+                raise RuntimeError(f"Failed at batch index {i+1}: {e}")
```
Comment (paquiteau) [inline src/mrinufft/operators/autodiff.py:212]:
```suggestion
            except Exception as e:
                raise RuntimeError(f"Failed at batch index {i+1}") from e
```

File: src/mrinufft/operators/base.py [hunk 3]
```diff
@@ -286,9 +308,20 @@ def make_autograd(self, wrt_data=True, wrt_traj=False):
         if not self.autograd_available:
             raise ValueError("Backend does not support auto-differentiation.")
 
-        from mrinufft.operators.autodiff import MRINufftAutoGrad
```
Comment (paquiteau) [inline src/mrinufft/operators/base.py:289]: You can do the two imports here

File: tests/helpers/__init__.py [hunk 4]
```diff
@@ -4,6 +4,7 @@
 from .factories import (
     kspace_from_op,
     image_from_op,
+    batchedSmpas_from_op,
```
Comment (paquiteau) [inline tests/helpers/__init__.py:7]: typo

File: tests/helpers/factories.py [hunk 5]
```diff
@@ -19,28 +19,42 @@
...
+    kspace = (
+        1j * np.random.randn(*batch_dim, operator.n_coils, operator.n_samples)
+    ).astype(operator.cpx_dtype)
+    kspace += np.random.randn(*batch_dim, operator.n_coils, operator.n_samples).astype(
         operator.cpx_dtype
     )
     return kspace
 
 
+def batchedSmpas_from_op(operator):
```
Comment (paquiteau) [inline tests/helpers/factories.py:46]: typo

File: tests/operators/test_autodiff.py
Comment (paquiteau) [inline tests/operators/test_autodiff.py:1]: The tests are quite long and complex to read, I would advise to refactor and split them between batched and non-batched version

File: src/mrinufft/operators/autodiff.py [hunk 6]
```diff
@@ -152,3 +152,94 @@ def samples(self, value):
     def __getattr__(self, name):
         """Forward all other attributes to the nufft_op."""
         return getattr(self.nufft_op, name)
+
+
+class BatchedNufftAutoGrad(MRINufftAutoGrad):
```
Comment (chaithyagr) [inline src/mrinufft/operators/autodiff.py:157]: I am curious, do we need a different class for this? Why not support it in the main class itself? Other option is just to add `batch_size` as an argument to make_autograd.

Comment (AsmaTANABEN) [reply src/mrinufft/operators/autodiff.py:157]: I believe separating them helps keep each class clean and easier to extend. My initial choice was to keep MRINufftAutoGrad as the base class without adding unnecessary complexity, but I'm open to merging the two classes. @paquiteau what do you think?

Comment (paquiteau) [reply src/mrinufft/operators/autodiff.py:157]: After some more though about it, I guess we could merge the two classes together (but keep things tidy with `_{adj}_op_batched` methods if needed) . However. it is not quite clear to me if the smaps are and/or need to be saved in context for this batched version to work

Comment (paquiteau) [reply src/mrinufft/operators/autodiff.py]: great ! A comment in the code could be added to explicit that :)

File: src/mrinufft/operators/autodiff.py [hunk 7]
```diff
@@ -109,9 +109,17 @@ class MRINufftAutoGrad(torch.nn.Module):
     Parameters
     ----------
     nufft_op: Classic Non differentiable MRI-NUFFT operator.
+
+    batch_size : int, default None
+        Number of batches to process simultaneously.
+        if Provided, the Nufft operator will support different data/smaps pairs
```
Comment (chaithyagr) [inline src/mrinufft/operators/autodiff.py:115]:
```suggestion
        if provided, the NUFFT operator will support different processing
        different k-space / image data and corresponding sensitivity map pairs.
        This is particularly useful for multi-contrast reconstructions
```

File: src/mrinufft/operators/autodiff.py [hunk 8]
```diff
@@ -109,9 +109,17 @@ class MRINufftAutoGrad(torch.nn.Module):
...
+            in a batched mode
+
     """
 
-    def __init__(self, nufft_op, wrt_data=True, wrt_traj=False):
+    # TODO Future improvements may include support for varying trajs across batches.
```
Comment (chaithyagr) [inline src/mrinufft/operators/autodiff.py:120]: Please remove this and make it an issue

File: src/mrinufft/operators/autodiff.py [hunk 9]
```diff
@@ -127,15 +135,59 @@ def __init__(self, nufft_op, wrt_data=True, wrt_traj=False):
             # used for update also.
             self._samples_torch = torch.Tensor(self.nufft_op.samples)
             self._samples_torch.requires_grad = True
+        self.batch_size = batch_size
+        self.paired_batch_mode = batch_size is not None
```
Comment (chaithyagr) [inline src/mrinufft/operators/autodiff.py:139]: You do have `self.batch_size` always with you, do you really need another boolean?

File: src/mrinufft/operators/autodiff.py [hunk 10]
```diff
@@ -152,3 +204,37 @@ def samples(self, value):
     def __getattr__(self, name):
         """Forward all other attributes to the nufft_op."""
         return getattr(self.nufft_op, name)
+
+    def _check_input_shape(self, *, imgs=None, ksps=None):
```
Comment (chaithyagr) [inline src/mrinufft/operators/autodiff.py:208]: This should be `_check_batch_size` right?

File: src/mrinufft/operators/autodiff.py [hunk 11]
```diff
@@ -127,15 +135,59 @@ def __init__(self, nufft_op, wrt_data=True, wrt_traj=False):
...
+
+    def adj_op_batched(self, batched_kspace, batched_smaps):
+        """Compute the adjoint batched_kspace -> batched_imgs."""
+        # NUFFT op is saved per batch element in ctx to ensure correct backpropagation.
+        self._check_input_shape(ksps=batched_kspace)
+        self._check_input_shape(imgs=batched_smaps)
+        batched_imgs = []
+        for i in range(self.batch_size):
+            try:
+                self.nufft_op.smaps = batched_smaps[i]
```
Comment (chaithyagr) [inline src/mrinufft/operators/autodiff.py:183]: Careful, you could be doing a lot of copies and throttling. But I am okay with it as alternative is just to make many NUFFT ops, which I expect user to make in any case.

File: src/mrinufft/operators/base.py [hunk 12]
```diff
@@ -50,7 +50,12 @@ def list_backends(available_only=False):
 
 
 def get_operator(
-    backend_name: str, wrt_data: bool = False, wrt_traj: bool = False, *args, **kwargs
+    backend_name: str,
+    wrt_data: bool = False,
+    wrt_traj: bool = False,
+    paired_batch_size: int = None,
```
Comment (chaithyagr) [inline src/mrinufft/operators/base.py:56]: If I understand correctly `paired_batch_size` is used only if we want to process for different smaps right? 

Can we rename in that case? `smap_batch_size` ? @paquiteau opinions?

File: tests/helpers/__init__.py [hunk 13]
```diff
@@ -4,6 +4,7 @@
 from .factories import (
     kspace_from_op,
     image_from_op,
+    batchedSmaps_from_op,
```
Comment (chaithyagr) [inline tests/helpers/__init__.py:7]:
```suggestion
    batched_smaps_from_op,
```
Lets follow the same naming conventions :)

File: tests/operators/test_autodiff.py [hunk 14]
```diff
@@ -109,30 +170,79 @@ def test_adjoint_and_grad(operator, interface):
...
     )
-
     # Check if nufft and ndft w.r.t trajectory are close in the backprop
     gradient_ndft_ktraj = torch.autograd.grad(
         loss_ndft, operator.samples, retain_graph=True
     )[0]
     gradient_nufft_ktraj = torch.autograd.grad(
         loss_nufft, operator.samples, retain_graph=True
     )[0]
+    # FIXME: atol=5e-1 is too loose?
```
Comment (chaithyagr) [inline tests/operators/test_autodiff.py:180]: Oh yes indeed. Lets discuss how much difference you have

Comment (chaithyagr) [reply tests/operators/test_autodiff.py:180]: It could be possible that you have wrong gradients also

**Review by chaithyagr — CHANGES_REQUESTED** [review]
Thank you for this updates. A bunch of comments and major changes from me.

File: src/mrinufft/operators/base.py (hunk 12 above)
Comment (paquiteau) [reply src/mrinufft/operators/base.py:56]: From my understanding in this "paired" mode, the batch size of smaps and data is the same. Therefore I don't think we need an extra integer, a simple boolean flag `paired_batch` (or `paired_batch_mode` fwiw) should be enough ?

File: src/mrinufft/operators/autodiff.py [hunk 15]
```diff
@@ -127,15 +159,118 @@ def __init__(self, nufft_op, wrt_data=True, wrt_traj=False):
...
+
+    def op(self, x, smaps=None, samples=None):
+        r"""Compute the forward image -> k-space.
+
+        Parameters
+        ----------
+        x: Tensor
+            Image data of shape ``(batch_size, nufft_op.n_batchs, (1 if smaps
+            else nufft_op.n_coils), *nufft_op.shape)``
+        smaps: Tensor, optional
```
Comment (chaithyagr) [inline src/mrinufft/operators/autodiff.py]: Why tensor? I think this isn't how others are documented right?

File: src/mrinufft/operators/autodiff.py [hunk 16]
```diff
@@ -127,15 +159,118 @@ def __init__(self, nufft_op, wrt_data=True, wrt_traj=False):
...
+            Sensitivity maps of shape ``(batch_size, nufft_op.n_coils, *nufft_op.shape)
+        samples: Tensor, optional
+            Samples for the batches of shape ``(batch_size, nufft_op.n_samples,
+            nufft_op.ndim)``
+
+        Returns
+        -------
+        Tensor:
```
Comment (chaithyagr) [inline src/mrinufft/operators/autodiff.py]: I see this at many places

File: src/mrinufft/operators/base.py [hunk 17]
```diff
@@ -559,9 +583,20 @@ def __repr__(self):
         )
 
     @classmethod
-    def with_autograd(cls, wrt_data=True, wrt_traj=False, *args, **kwargs):
+    def with_autograd(
+        cls,
+        wrt_data=True,
+        wrt_traj=False,
+        paired_batch_size=None,
```
Comment (chaithyagr) [inline src/mrinufft/operators/base.py]: Wait here it's banned something else, I'm getting confused

File: tests/helpers/__init__.py (hunk 4 above)
Comment (chaithyagr) [reply tests/helpers/__init__.py]: I see this still?

**Review by chaithyagr — CHANGES_REQUESTED** [review]
Thanku you for taking it forward. Some old comments still aren't addressed and i found some major changes required in docs. Finally, do we have tests working to test this?
